src/message_dao.py

```diff
@@ -12,27 +12,141 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import Dict, Optional
-from message import Message
+import json
+import os
+import subprocess
 
+from functools import lru_cache
+from typing import Dict, List, Optional
+
+from dotenv import load_dotenv
+from google.cloud.sql.connector import Connector
+from message import lore_link, Message, parse_message_from_str
+
+load_dotenv()
 EPOCH_HASH = 'ae9e7be4a03765456fe38287533e6446e8bbc93c'
 
 class MessageDao(object):
+    def __init__(self, archive_path: str) -> None:
+        """ Creates a connection as well as two tables: Messages and States.
+        Message stores the messages we've uploaded and States is a key-value
+        store which tracks things like 'last_hash', the last Lore git commit
+        we've processed."""
+        self._initialize_connection()
+        self._initialize_tables()
+        self.archive_path = archive_path
+
+    def _initialize_connection(self) -> None:
+        connector = Connector()
+        self.connection = connector.connect(
+            os.environ.get("HOST"),
+            "pymysql",
+            user = os.environ.get("USER"),
+            password = os.environ.get("PASSWORD")
+        )
+    
+    def _initialize_tables(self) -> None:
+        db_name = os.environ.get("DB")
+        if not db_name:
+            raise Exception("Missing environment variable for name of database.")
+        with self.connection.cursor() as cursor:
+            cursor.execute("CREATE DATABASE IF NOT EXISTS " + db_name)
+            self.connection.select_db(db_name)
+            # Mapping from message id to message
+            cursor.execute(
+                "CREATE TABLE IF NOT EXISTS Messages"
+                "(message_id VARCHAR(255) NOT NULL,"
+                "reduced_subject VARCHAR(255) NOT NULL,"
+                "from_ VARCHAR(255) NOT NULL,"
+                "in_reply_to VARCHAR(255),"
+                "archive_hash VARCHAR(255) NOT NULL,"
+                "change_id VARCHAR(255),"
+                "lore_link VARCHAR(255),"
+                "PRIMARY KEY (message_id))"
+            )
+            # Mapping from name of state attribute to state
+            cursor.execute(
+                "CREATE TABLE IF NOT EXISTS States"
+                "(state_name VARCHAR(255) NOT NULL,"
+                "value VARCHAR(255) NOT NULL,"
+                "PRIMARY KEY (state_name))"
+            )
+        self.connection.commit()
+
+    def store(self, message: Message) -> None:
+        # Use only the portion of the subject that comes after the brackets that indicate patch information
+        reduced_subject = message.subject.partition("] ")[2].lower()
+        link = lore_link(message.id)
```

Comment:
as noted in the other thread, we don't need to store the lore_link, it's just a concatenation of a constant and the msg_id.

Reply:
As Brendan mentioned, the idea was to have an easy to access link to the raw data itself. Although all we need is the message id, storing the link does allow for just a simple click. What are your thoughts?

Reply:
I don't have any strong preference.
Let's keep it if you and Brendan agree.

I don't personally think it'd be too useful, but it's only adding a small number of bytes.

Reply:
I am still leaning towards keeping the whole lore link: If we plan on opening this up to other email lists in the future, then we should store the list here as well. But then if we keep both the list name as well as the message id, then we are only going to save 23 chars. Now, I know this is 23 times "a lot", which is a lot; however, I think it also makes it meaningfully easier to look up the patch on lore then having to copy and paste and type 23 chars into chrome. For reference, a typical full lore link will be about 90 bytes.

I fully expect that we will spend non-trivial amounts of time on our DB researching improvements to the lkml-gerrit-bridge. Even if it is only interns in the summer, I think 1 intern next summer spending 10 hours messing around on the DB would make making the lore-link copy-and-paste-able worthwhile.

Now, there are other compromises we could make: We could just cut the "https://" off of the link, reducing the overhead of copy-and-paste-ability to 15 bytes. We could also make a go link ("go/llore" is available) to reduce the pain of typing a prefix before copying and pasting, but it's still enough characters that I would find typing it out to be annoying.